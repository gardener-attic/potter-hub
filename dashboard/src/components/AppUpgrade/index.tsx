import AppUpgrade from "./AppUpgrade";

export default AppUpgrade;
