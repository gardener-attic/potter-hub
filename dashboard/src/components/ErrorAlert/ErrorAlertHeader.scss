.error__icon {
  display: inline-block;
  background: #c7c9c8;
  vertical-align: bottom;
  border-radius: 50%;
  min-height: 30px;
  min-width: 30px;
}

.error__icon > svg {
  margin: auto;
  left: 3px;
  top: 3px;
  color: #ffffff;
  position: relative;
}
