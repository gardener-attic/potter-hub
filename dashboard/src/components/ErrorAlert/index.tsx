import ErrorAlertHeader from "./ErrorAlertHeader";
import ErrorSelector from "./ErrorSelector";
import MessageAlert from "./MessageAlert";
import NotFoundErrorAlert from "./NotFoundErrorAlert";
import PermissionsErrorAlert from "./PermissionsErrorAlert";
import ServiceBrokersNotFoundAlert from "./ServiceBrokersNotFoundAlert";
import ServiceCatalogNotInstalledAlert from "./ServiceCatalogNotInstalledAlert";
import UnexpectedErrorAlert from "./UnexpectedErrorAlert";

export {
  ErrorAlertHeader,
  ErrorSelector,
  MessageAlert,
  NotFoundErrorAlert,
  PermissionsErrorAlert,
  ServiceBrokersNotFoundAlert,
  ServiceCatalogNotInstalledAlert,
  UnexpectedErrorAlert,
};
