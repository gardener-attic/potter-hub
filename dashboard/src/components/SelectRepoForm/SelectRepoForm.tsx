import * as React from "react";

import { IAppRepository } from "../../shared/types";
import LoadingWrapper from "../LoadingWrapper";

import { ErrorSelector } from "../ErrorAlert";

interface ISelectRepoFormProps {
  kubeappsNamespace: string;
  error: Error | undefined;
  repo: IAppRepository;
  repos: IAppRepository[];
  chartName: string;
  checkChart: (repo: string, chartName: string) => any;
}

interface ISelectRepoFormState {
  repo: string;
}

class SelectRepoForm extends React.Component<ISelectRepoFormProps, ISelectRepoFormState> {
  public state: ISelectRepoFormState = {
    repo:
      this.props.repo.metadata && this.props.repo.metadata.name
        ? this.props.repo.metadata.name
        : "",
  };

  public render() {
    return (
      <LoadingWrapper loaded={this.props.repos.length > 0}>
        <div className="container margin-normal">
          <div className="col-8">
            {this.props.error && (
              <ErrorSelector
                error={this.props.error}
                resource={`Chart ${this.state.repo}/${this.props.chartName}`}
              />
            )}
          </div>
          <div className="col-12">
            <h2>Select the source repository of {this.props.chartName}</h2>
          </div>
          <div className="col-8">
            <label htmlFor="chartRepoName">Chart Repository Name *</label>
            <select
              id="chartRepoName"
              onChange={this.handleChartRepoNameChange}
              value={this.state.repo}
              required={true}
            >
              {!this.state.repo && <option key="" value="" />}
              {this.props.repos.map(r => (
                <option key={r.metadata.name} value={r.metadata.name}>
                  {r.metadata.name} ({this.getRepoURL(r.metadata.name)})
                </option>
              ))}
            </select>
          </div>
          <div>
            <p>
              {" "}
              * If the repository containing {this.props.chartName} is not in the list add it{" "}
              <a href="/config/repos"> here </a>.{" "}
            </p>
          </div>
        </div>
      </LoadingWrapper>
    );
  }

  public handleChartRepoNameChange = async (e: React.ChangeEvent<HTMLSelectElement>) => {
    this.props.checkChart(e.currentTarget.value, this.props.chartName);
    this.setState({ repo: e.currentTarget.value });
  };

  private getRepoURL = (name: string) => {
    let res = "";
    this.props.repos.forEach(r => {
      if (r.metadata.name === name && r.spec) {
        res = r.spec.url;
      }
    });
    return res;
  };
}

export default SelectRepoForm;
