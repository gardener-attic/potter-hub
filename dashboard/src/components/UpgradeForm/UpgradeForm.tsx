import { RouterAction } from "connected-react-router";
import * as jsonpatch from "fast-json-patch";
import { JSONSchema4 } from "json-schema";
import * as React from "react";
import * as YAML from "yaml";

import { deleteValue, setValue } from "../../shared/schema";
import { IChartState, IChartVersion } from "../../shared/types";
import DeploymentFormBody from "../DeploymentFormBody/DeploymentFormBody";
import { ErrorSelector } from "../ErrorAlert";
import LoadingWrapper from "../LoadingWrapper";

export interface IUpgradeFormProps {
  appCurrentVersion: string;
  appCurrentValues?: string;
  chartName: string;
  namespace: string;
  releaseName: string;
  repo: string;
  error: Error | undefined;
  selected: IChartState["selected"];
  deployed: IChartState["deployed"];
  upgradeApp: (
    version: IChartVersion,
    releaseName: string,
    namespace: string,
    values?: string,
    schema?: JSONSchema4,
  ) => Promise<boolean>;
  push: (location: string) => RouterAction;
  goBack: () => RouterAction;
  fetchChartVersions: (id: string) => Promise<IChartVersion[] | undefined>;
  getChartVersion: (id: string, chartVersion: string) => void;
}

interface IUpgradeFormState {
  appValues: string;
  valuesModified: boolean;
  isDeploying: boolean;
  modifications?: jsonpatch.Operation[];
  deployedValues?: string;
}

class UpgradeForm extends React.Component<IUpgradeFormProps, IUpgradeFormState> {
  public state: IUpgradeFormState = {
    appValues: this.props.appCurrentValues || "",
    isDeploying: false,
    valuesModified: false,
  };

  public componentDidMount() {
    const chartID = `${this.props.repo}/${this.props.chartName}`;
    this.props.fetchChartVersions(chartID);
  }

  public componentDidUpdate = (prevProps: IUpgradeFormProps) => {
    let modifications = this.state.modifications;
    // applyModifications is an expensive operatior, that's why it's only defined within
    // the componentDidUpdate scope
    const applyModifications = (mods: jsonpatch.Operation[], appValues: string) => {
      // And we add any possible change made to the original version
      if (mods.length) {
        mods.forEach(modification => {
          if (modification.op === "remove") {
            appValues = deleteValue(appValues, modification.path);
          } else {
            // Transform the modification as a ReplaceOperation to read its value
            const value = (modification as jsonpatch.ReplaceOperation<any>).value;
            appValues = setValue(appValues, modification.path, value);
          }
        });
      }
      return appValues;
    };

    if (this.props.deployed.values && !modifications) {
      // Calculate modifications from the default values
      const defaultValuesObj = YAML.parse(this.props.deployed.values);
      const deployedValuesObj = YAML.parse(this.props.appCurrentValues || "");
      modifications = jsonpatch.compare(defaultValuesObj, deployedValuesObj);
      const values = applyModifications(modifications, this.props.deployed.values);
      this.setState({ modifications });
      this.setState({ appValues: values, deployedValues: values });
    }

    if (prevProps.selected.version !== this.props.selected.version && !this.state.valuesModified) {
      // Apply modifications to the new selected version
      const appValues = modifications
        ? applyModifications(modifications, this.props.selected.values || "")
        : this.props.selected.values || "";
      this.setState({ appValues });
    }
  };

  public render() {
    const { namespace, releaseName, error, selected } = this.props;
    if (error) {
      return (
        <ErrorSelector error={error} namespace={namespace} action="update" resource={releaseName} />
      );
    }
    if (selected.versions.length === 0) {
      return <LoadingWrapper />;
    }
    const chartID = `${this.props.repo}/${this.props.chartName}`;
    return (
      <form className="container padding-b-bigger" onSubmit={this.handleDeploy}>
        <div className="row">
          <div className="col-12">
            <h2>{`${this.props.releaseName} (${chartID})`}</h2>
          </div>
          <div className="col-8">
            <DeploymentFormBody
              chartID={chartID}
              chartVersion={this.props.appCurrentVersion}
              deployedValues={this.state.deployedValues}
              namespace={this.props.namespace}
              releaseVersion={this.props.appCurrentVersion}
              selected={this.props.selected}
              push={this.props.push}
              goBack={this.props.goBack}
              getChartVersion={this.props.getChartVersion}
              setValues={this.handleValuesChange}
              appValues={this.state.appValues}
              setValuesModified={this.setValuesModified}
            />
          </div>
        </div>
      </form>
    );
  }

  public setValuesModified = () => {
    this.setState({ valuesModified: true });
  };

  public handleValuesChange = (value: string) => {
    this.setState({ appValues: value });
  };

  public handleDeploy = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const { selected, push, upgradeApp, releaseName, namespace } = this.props;
    const { appValues } = this.state;

    this.setState({ isDeploying: true });
    if (selected.version) {
      const deployed = await upgradeApp(
        selected.version,
        releaseName,
        namespace,
        appValues,
        selected.schema,
      );
      this.setState({ isDeploying: false });
      if (deployed) {
        push(`/apps/ns/${namespace}/${releaseName}`);
      }
    }
  };
}

export default UpgradeForm;
