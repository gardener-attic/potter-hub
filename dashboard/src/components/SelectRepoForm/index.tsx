import SelectRepoForm from "./SelectRepoForm";

export default SelectRepoForm;
