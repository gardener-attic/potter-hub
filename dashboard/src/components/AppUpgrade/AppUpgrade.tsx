import * as React from "react";

import { RouterAction } from "connected-react-router";
import { JSONSchema4 } from "json-schema";
import { Link } from "react-router-dom";
import {
  IAppRepository,
  IChartState,
  IChartVersion,
  IRBACRole,
  IRelease,
} from "../../shared/types";
import { ErrorSelector, MessageAlert } from "../ErrorAlert";
import LoadingWrapper from "../LoadingWrapper";
import SelectRepoForm from "../SelectRepoForm";
import UpgradeForm from "../UpgradeForm";

interface IAppUpgradeProps {
  app?: IRelease;
  isFetching: boolean;
  error: Error | undefined;
  repoError: Error | undefined;
  kubeappsNamespace: string;
  namespace: string;
  releaseName: string;
  version: string;
  repos: IAppRepository[];
  repo: IAppRepository;
  selected: IChartState["selected"];
  deployed: IChartState["deployed"];
  upgradeApp: (
    version: IChartVersion,
    releaseName: string,
    namespace: string,
    values?: string,
    schema?: JSONSchema4,
  ) => Promise<boolean>;
  clearRepo: () => void;
  checkChart: (repo: string, chartName: string) => void;
  fetchChartVersions: (id: string) => Promise<IChartVersion[] | undefined>;
  getAppWithUpdateInfo: (releaseName: string, namespace: string) => void;
  getChartVersion: (id: string, chartVersion: string) => void;
  getDeployedChartVersion: (id: string, chartVersion: string) => void;
  push: (location: string) => RouterAction;
  goBack: () => RouterAction;
  fetchRepositories: () => void;
}

interface IAppUpgradeState {
  repo?: IAppRepository;
}

class AppUpgrade extends React.Component<IAppUpgradeProps, IAppUpgradeState> {
  public state: IAppUpgradeState = {};

  public componentDidMount() {
    const { releaseName, getAppWithUpdateInfo, namespace, fetchRepositories } = this.props;
    getAppWithUpdateInfo(releaseName, namespace);
    fetchRepositories();
  }

  public componentDidUpdate(prevProps: IAppUpgradeProps) {
    const { repos, app } = this.props;
    const prevRepo = this.state.repo;
    let repo = this.state.repo;
    // Retrieve the current repo
    if (!repo) {
      repo = this.props.repo;
      if (repo && repo.metadata) {
        // If the repository comes from the properties, use it
        this.setState({ repo });
      } else {
        // If there is update info about the app we can automatically chose the repository
        // with the latest version
        if (app && app.updateInfo) {
          const repoWithLatest = repos.find(
            r => r.metadata.name === (app.updateInfo && app.updateInfo.repository.name),
          );
          if (repoWithLatest) {
            this.setState({ repo: repoWithLatest });
            repo = repoWithLatest;
          }
        }
      }
    }

    if (app) {
      const { chart } = app;
      if (
        chart &&
        chart.metadata &&
        chart.metadata.name &&
        chart.metadata.version &&
        repo &&
        repo.metadata &&
        repo.metadata.name &&
        (prevProps.app !== app || prevRepo !== repo)
      ) {
        const chartID = `${repo.metadata.name}/${chart.metadata.name}`;
        this.props.getDeployedChartVersion(chartID, chart.metadata.version);
      }
    }
  }

  public render() {
    const { app, repos, error, namespace, releaseName, repoError, isFetching } = this.props;
    const { repo } = this.state;
    if (isFetching) {
      return <LoadingWrapper />;
    }
    if (
      !repos ||
      repos.length === 0 ||
      !app ||
      !app.chart ||
      !app.chart.metadata ||
      !app.chart.metadata.name ||
      !app.chart.metadata.version
    ) {
      if (error) {
        return (
          <ErrorSelector
            error={error}
            namespace={namespace}
            action="update"
            defaultRequiredRBACRoles={{ update: this.requiredRBACRoles() }}
            resource={releaseName}
          />
        );
      } else if (repoError) {
        return (
          <ErrorSelector
            error={repoError}
            namespace={this.props.kubeappsNamespace}
            action="view"
            defaultRequiredRBACRoles={{ view: this.requiredRBACRoles() }}
            resource="App Repositories"
          />
        );
      } else if (repos.length === 0) {
        return (
          <MessageAlert
            level={"warning"}
            children={
              <div>
                <h5>Chart repositories not found.</h5>
                Manage your Helm chart repositories in Kubeapps by visiting the{" "}
                <Link to={"/config/repos"}>App repositories configuration</Link> page.
              </div>
            }
          />
        );
      } else {
        return (
          <ErrorSelector
            error={new Error("Unable to obtain the required information to upgrade")}
            resource={releaseName}
          />
        );
      }
    }
    if (!repo || !repo.metadata) {
      return (
        <div>
          <SelectRepoForm
            {...this.props}
            error={this.props.repoError}
            chartName={app.chart.metadata.name}
          />
        </div>
      );
    }
    return (
      <div>
        <UpgradeForm
          {...this.props}
          appCurrentVersion={app.chart.metadata.version}
          appCurrentValues={(app.config && app.config.raw) || ""}
          chartName={app.chart.metadata.name}
          repo={repo.metadata.name}
        />
      </div>
    );
  }

  private requiredRBACRoles(): IRBACRole[] {
    return [
      {
        apiGroup: "kubeapps.com",
        namespace: this.props.kubeappsNamespace,
        resource: "apprepositories",
        verbs: ["get"],
      },
    ];
  }
}

export default AppUpgrade;
