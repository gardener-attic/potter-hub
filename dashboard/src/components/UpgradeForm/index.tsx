import UpgradeForm from "./UpgradeForm";

export default UpgradeForm;
