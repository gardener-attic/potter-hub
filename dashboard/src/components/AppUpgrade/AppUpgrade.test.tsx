import { mount, shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";

import { hapi } from "shared/hapi/release";
import itBehavesLike from "../../shared/specs";
import { ForbiddenError, IAppRepository, IChartState, IRelease } from "../../shared/types";
import { ErrorSelector, MessageAlert, PermissionsErrorAlert } from "../ErrorAlert";
import ErrorPageHeader from "../ErrorAlert/ErrorAlertHeader";
import SelectRepoForm from "../SelectRepoForm";
import UpgradeForm from "../UpgradeForm";
import AppUpgrade from "./AppUpgrade";

const defaultProps = {
  app: {} as hapi.release.Release,
  isFetching: false,
  checkChart: jest.fn(),
  clearRepo: jest.fn(),
  error: undefined,
  fetchChartVersions: jest.fn(),
  fetchRepositories: jest.fn(),
  getAppWithUpdateInfo: jest.fn(),
  getChartVersion: jest.fn(),
  deployed: {} as IChartState["deployed"],
  getDeployedChartVersion: jest.fn(),
  kubeappsNamespace: "kubeapps",
  namespace: "default",
  push: jest.fn(),
  goBack: jest.fn(),
  releaseName: "foo",
  repo: {} as IAppRepository,
  repoError: undefined,
  repos: [],
  selected: {} as IChartState["selected"],
  upgradeApp: jest.fn(),
  version: "1.0.0",
};

beforeEach(() => {
  jest.resetAllMocks();
});

itBehavesLike("aLoadingComponent", {
  component: AppUpgrade,
  props: { ...defaultProps, isFetching: true },
});

it("renders the repo selection form if not introduced", () => {
  const wrapper = shallow(
    <AppUpgrade
      {...defaultProps}
      app={
        {
          chart: {
            metadata: {
              name: "bar",
              version: "1.0.0",
            },
          },
          name: "foo",
        } as IRelease
      }
      repos={[
        {
          metadata: { name: "stable" },
        } as IAppRepository,
      ]}
    />,
  );
  expect(wrapper.find(SelectRepoForm)).toExist();
  expect(wrapper.find(ErrorSelector)).not.toExist();
  expect(wrapper.find(UpgradeForm)).not.toExist();
  expect(wrapper).toMatchSnapshot();
});

context("when an error exists", () => {
  it("renders a generic error message", () => {
    const repo = {
      metadata: { name: "stable" },
    } as IAppRepository;
    const wrapper = shallow(
      <AppUpgrade
        {...defaultProps}
        error={new Error("foo doesn't exists")}
        repos={[repo]}
        repo={repo}
      />,
    );

    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper.find(SelectRepoForm)).not.toExist();
    expect(wrapper.find(UpgradeForm)).not.toExist();

    expect(wrapper.html()).toContain("Sorry! Something went wrong.");
    expect(wrapper).toMatchSnapshot();
  });

  it("renders a forbidden message", () => {
    const repo = {
      metadata: { name: "stable" },
    } as IAppRepository;
    const role = {
      apiGroup: "kubeapps.com",
      namespace: "kubeapps",
      resource: "apprepositories",
      verbs: ["get"],
    };
    const wrapper = mount(
      <AppUpgrade
        {...defaultProps}
        error={new ForbiddenError(JSON.stringify([role]))}
        repos={[repo]}
        repo={repo}
      />,
    );

    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper.find(SelectRepoForm)).not.toExist();
    expect(wrapper.find(UpgradeForm)).not.toExist();

    expect(wrapper.find(ErrorPageHeader).text()).toContain(
      "You don't have sufficient permissions to update foo in the default namespace",
    );
    expect(wrapper.find(PermissionsErrorAlert).prop("roles")[0]).toMatchObject(role);
  });

  it("renders a forbidden message for the repositories", () => {
    const role = {
      apiGroup: "kubeapps.com",
      namespace: "kubeapps",
      resource: "apprepositories",
      verbs: ["get"],
    };
    const wrapper = mount(<AppUpgrade {...defaultProps} repoError={new ForbiddenError()} />);

    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper.find(SelectRepoForm)).not.toExist();
    expect(wrapper.find(UpgradeForm)).not.toExist();

    expect(wrapper.find(ErrorPageHeader).text()).toContain(
      "You don't have sufficient permissions to view App Repositories in the kubeapps namespace",
    );
    expect(wrapper.find(PermissionsErrorAlert).prop("roles")[0]).toMatchObject(role);
  });

  it("renders a warning message if there are no repositories", () => {
    const wrapper = shallow(<AppUpgrade {...defaultProps} />);

    expect(wrapper.find(MessageAlert)).toExist();
    expect(wrapper.find(SelectRepoForm)).not.toExist();
    expect(wrapper.find(UpgradeForm)).not.toExist();

    expect(
      wrapper
        .find(MessageAlert)
        .children()
        .text(),
    ).toContain("Chart repositories not found");
  });

  it("renders an error message if the app information is missing some metadata", () => {
    const repo = {
      metadata: { name: "stable" },
    } as IAppRepository;
    const wrapper = mount(
      <AppUpgrade
        {...defaultProps}
        repos={[repo]}
        app={
          {
            chart: {
              metadata: {},
            },
            name: "foo",
          } as IRelease
        }
      />,
    );

    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper.find(SelectRepoForm)).not.toExist();
    expect(wrapper.find(UpgradeForm)).not.toExist();

    expect(wrapper.find(ErrorSelector).text()).toContain(
      "Unable to obtain the required information to upgrade",
    );
  });
});

it("renders the upgrade form when the repo is available", () => {
  const repo = {
    metadata: { name: "stable" },
  } as IAppRepository;
  const wrapper = shallow(
    <AppUpgrade
      {...defaultProps}
      app={
        {
          chart: {
            metadata: {
              name: "bar",
              version: "1.0.0",
            },
          },
          name: "foo",
        } as IRelease
      }
      repos={[repo]}
    />,
  );
  wrapper.setProps({ repo });
  expect(wrapper.find(UpgradeForm)).toExist();
  expect(wrapper.find(ErrorSelector)).not.toExist();
  expect(wrapper.find(SelectRepoForm)).not.toExist();
  expect(wrapper).toMatchSnapshot();
});

it("skips the repo selection form if the app contains upgrade info", () => {
  const repo = {
    metadata: { name: "stable" },
  } as IAppRepository;
  const app = {
    chart: {
      metadata: {
        name: "bar",
        version: "1.0.0",
      },
    },
    name: "foo",
    updateInfo: {
      upToDate: true,
      chartLatestVersion: "1.1.0",
      appLatestVersion: "1.1.0",
      repository: { name: "stable", url: "" },
    },
  } as IRelease;
  const wrapper = shallow(<AppUpgrade {...defaultProps} repos={[repo]} />);
  wrapper.setProps({ app });
  expect(wrapper.find(UpgradeForm)).toExist();
  expect(wrapper.find(ErrorSelector)).not.toExist();
  expect(wrapper.find(SelectRepoForm)).not.toExist();
  expect(wrapper).toMatchSnapshot();
});

describe("when receiving new props", () => {
  it("should set the source repository in the state if present as a property", () => {
    const repo = { metadata: { name: "stable" } };
    const wrapper = shallow(<AppUpgrade {...defaultProps} />);
    wrapper.setProps({ repo });
    expect(wrapper.state("repo")).toEqual(repo);
  });

  it("should request the deployed chart when the app and repo are populated", () => {
    const repo = { metadata: { name: "stable" } };
    const app = {
      chart: {
        metadata: {
          name: "bar",
          version: "1.0.0",
        },
      },
    } as IRelease;
    const getDeployedChartVersion = jest.fn();
    const wrapper = shallow(
      <AppUpgrade {...defaultProps} getDeployedChartVersion={getDeployedChartVersion} />,
    );
    wrapper.setProps({ repo, app });
    expect(getDeployedChartVersion).toHaveBeenCalledWith("stable/bar", "1.0.0");
  });

  it("should request the deployed chart when the repo is populated later", () => {
    const repo = { metadata: { name: "stable" } };
    const app = {
      chart: {
        metadata: {
          name: "bar",
          version: "1.0.0",
        },
      },
    } as IRelease;
    const getDeployedChartVersion = jest.fn();
    const wrapper = shallow(
      <AppUpgrade {...defaultProps} app={app} getDeployedChartVersion={getDeployedChartVersion} />,
    );
    expect(getDeployedChartVersion).not.toHaveBeenCalled();
    wrapper.setProps({ repo });
    expect(getDeployedChartVersion).toHaveBeenCalledWith("stable/bar", "1.0.0");
  });

  it("a new app should re-trigger the deployed chart retrieval", () => {
    const repo = { metadata: { name: "stable" } };
    const app = {
      chart: {
        metadata: {
          name: "bar",
          version: "1.0.0",
        },
      },
    } as IRelease;
    const getDeployedChartVersion = jest.fn();
    const wrapper = shallow(
      <AppUpgrade {...defaultProps} getDeployedChartVersion={getDeployedChartVersion} />,
    );
    wrapper.setProps({ repo, app });
    expect(getDeployedChartVersion).toHaveBeenCalledWith("stable/bar", "1.0.0");

    const app2 = {
      chart: {
        metadata: {
          name: "foobar",
          version: "1.0.0",
        },
      },
    } as IRelease;
    wrapper.setProps({ app: app2 });
    expect(getDeployedChartVersion).toHaveBeenCalledWith("stable/foobar", "1.0.0");
  });
});
