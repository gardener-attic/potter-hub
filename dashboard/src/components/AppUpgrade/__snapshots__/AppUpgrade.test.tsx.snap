// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
  type={0}
/>
`;

exports[`renders the repo selection form if not introduced 1`] = `
<div>
  <SelectRepoForm
    app={
      Object {
        "chart": Object {
          "metadata": Object {
            "name": "bar",
            "version": "1.0.0",
          },
        },
        "name": "foo",
      }
    }
    chartName="bar"
    checkChart={[MockFunction]}
    clearRepo={[MockFunction]}
    deployed={Object {}}
    fetchChartVersions={[MockFunction]}
    fetchRepositories={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    getAppWithUpdateInfo={
      [MockFunction] {
        "calls": Array [
          Array [
            "foo",
            "default",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    getChartVersion={[MockFunction]}
    getDeployedChartVersion={[MockFunction]}
    goBack={[MockFunction]}
    isFetching={false}
    kubeappsNamespace="kubeapps"
    namespace="default"
    push={[MockFunction]}
    releaseName="foo"
    repo={Object {}}
    repos={
      Array [
        Object {
          "metadata": Object {
            "name": "stable",
          },
        },
      ]
    }
    selected={Object {}}
    upgradeApp={[MockFunction]}
    version="1.0.0"
  />
</div>
`;

exports[`renders the upgrade form when the repo is available 1`] = `
<div>
  <UpgradeForm
    app={
      Object {
        "chart": Object {
          "metadata": Object {
            "name": "bar",
            "version": "1.0.0",
          },
        },
        "name": "foo",
      }
    }
    appCurrentValues=""
    appCurrentVersion="1.0.0"
    chartName="bar"
    checkChart={[MockFunction]}
    clearRepo={[MockFunction]}
    deployed={Object {}}
    fetchChartVersions={[MockFunction]}
    fetchRepositories={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    getAppWithUpdateInfo={
      [MockFunction] {
        "calls": Array [
          Array [
            "foo",
            "default",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    getChartVersion={[MockFunction]}
    getDeployedChartVersion={
      [MockFunction] {
        "calls": Array [
          Array [
            "stable/bar",
            "1.0.0",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    goBack={[MockFunction]}
    isFetching={false}
    kubeappsNamespace="kubeapps"
    namespace="default"
    push={[MockFunction]}
    releaseName="foo"
    repo="stable"
    repos={
      Array [
        Object {
          "metadata": Object {
            "name": "stable",
          },
        },
      ]
    }
    selected={Object {}}
    upgradeApp={[MockFunction]}
    version="1.0.0"
  />
</div>
`;

exports[`skips the repo selection form if the app contains upgrade info 1`] = `
<div>
  <UpgradeForm
    app={
      Object {
        "chart": Object {
          "metadata": Object {
            "name": "bar",
            "version": "1.0.0",
          },
        },
        "name": "foo",
        "updateInfo": Object {
          "appLatestVersion": "1.1.0",
          "chartLatestVersion": "1.1.0",
          "repository": Object {
            "name": "stable",
            "url": "",
          },
          "upToDate": true,
        },
      }
    }
    appCurrentValues=""
    appCurrentVersion="1.0.0"
    chartName="bar"
    checkChart={[MockFunction]}
    clearRepo={[MockFunction]}
    deployed={Object {}}
    fetchChartVersions={[MockFunction]}
    fetchRepositories={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    getAppWithUpdateInfo={
      [MockFunction] {
        "calls": Array [
          Array [
            "foo",
            "default",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    getChartVersion={[MockFunction]}
    getDeployedChartVersion={
      [MockFunction] {
        "calls": Array [
          Array [
            "stable/bar",
            "1.0.0",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    goBack={[MockFunction]}
    isFetching={false}
    kubeappsNamespace="kubeapps"
    namespace="default"
    push={[MockFunction]}
    releaseName="foo"
    repo="stable"
    repos={
      Array [
        Object {
          "metadata": Object {
            "name": "stable",
          },
        },
      ]
    }
    selected={Object {}}
    upgradeApp={[MockFunction]}
    version="1.0.0"
  />
</div>
`;

exports[`when an error exists renders a generic error message 1`] = `
<ErrorSelector
  action="update"
  defaultRequiredRBACRoles={
    Object {
      "update": Array [
        Object {
          "apiGroup": "kubeapps.com",
          "namespace": "kubeapps",
          "resource": "apprepositories",
          "verbs": Array [
            "get",
          ],
        },
      ],
    }
  }
  error={[Error: foo doesn't exists]}
  namespace="default"
  resource="foo"
/>
`;
